File: python3/008. String to Integer (atoi).py
# 8. String to Integer (atoi)
# Accepted 56ms

class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        str = str.lstrip()
        strLen = len(str)
        
        if strLen == 0 or (str[0] != '+' and str[0] != '-' and (str[0] < '0' or str[0] > '9')):
            return 0
        
        if str[0] == '+':
            isNeg = False
            startIdx = 1
        elif str[0] == '-':
            isNeg = True
            startIdx = 1
        else:
            isNeg = False
            startIdx = 0
        
        num = 0
        for c in str[startIdx:]:
            if c < '0' or c > '9': break
            num = (num * 10) + (ord(c) - 48)
        
        if isNeg: num = -num
        if num < -2147483648: num = -2147483648
        if num > 2147483647: num = 2147483647
        
        return num

File: src/go/008. String to Integer (atoi).go
// 8. String to Integer (atoi)
// Accepted 4ms

func myAtoi(str string) int {
    str = strings.TrimSpace(str)
    if len(str) == 0 || (str[0] != '+' && str[0] != '-' && (str[0] < '0' || str[0] > '9')) { return 0 }
    
    var (
        num int64 = 0
        startIdx int = 0
        isNeg = false
    )
    
    switch str[0] {
        case '-': isNeg = true; fallthrough
        case '+': startIdx = 1
    }
    
    for _, c := range str[startIdx:] {
        if c < '0' || c > '9' { break }
        
        num = (num * 10) + int64(c - '0')
        if (isNeg && (-num) < int64(math.MinInt32)) || (!isNeg && num > int64(math.MaxInt32)) { break }
    }
    
    if isNeg { num = -num }
    if num < int64(math.MinInt32) { num = int64(math.MinInt32) }
    if num > int64(math.MaxInt32) { num = int64(math.MaxInt32) }
    
    return int(num)
}
